lsp/nls/src/requests/completion.rs

```diff
@@ -1,16 +1,99 @@
 use codespan::ByteIndex;
 use codespan_lsp::position_to_byte_index;
 use log::debug;
-use lsp_server::{RequestId, Response, ResponseError};
+use lsp_server::{ErrorCode, RequestId, Response, ResponseError};
 use lsp_types::{CompletionItem, CompletionParams};
+use nickel_lang::{
+    identifier::Ident,
+    term::{Contract, MetaValue},
+    types::{AbsType, RowIteratorItem, Types},
+};
 use serde_json::Value;
 
 use crate::{
-    linearization::interface::TermKind,
+    linearization::{
+        interface::{TermKind, UsageState, ValueState},
+        LinearizationItem,
+    },
     server::Server,
     trace::{Enrich, Trace},
 };
 
+/// Find record fields for an item with the specified id.
+fn find_record_fields(
+    linearization: &Vec<LinearizationItem<Types>>,
+    id: usize,
+) -> Option<Vec<Ident>> {
+    linearization.iter().find_map(|item| match item.kind {
+        TermKind::Record(ref fields) if item.id == id => {
+            Some(fields.keys().map(|ident| ident.clone()).collect())
+        }
+        TermKind::Declaration(_, _, ValueState::Known(new_id))
+        | TermKind::Usage(UsageState::Resolved(new_id))
+            if item.id == id =>
+        {
+            find_record_fields(linearization, new_id)
+        }
+        _ => None,
+    })
+}
+
+/// Find a record contract of the item with the specified id.
+fn find_record_contract(
+    linearization: &Vec<LinearizationItem<Types>>,
+    id: usize,
+) -> Option<Contract> {
+    linearization.iter().find_map(|item| match &item.meta {
+        Some(MetaValue { contracts, .. }) if item.id == id => {
+            contracts.iter().find_map(|contract| {
+                if let Types(AbsType::Record(..)) = contract.types {
+                    Some(contract.clone())
+                } else {
+                    None
+                }
+            })
+        }
+        _ if item.id == id => match item.kind {
+            TermKind::Declaration(_, _, ValueState::Known(new_id))
+            | TermKind::Usage(UsageState::Resolved(new_id)) => {
+                find_record_contract(&linearization, new_id)
+            }
+            _ => None,
+        },
+        _ => None,
+    })
+}
+
+/// Extract identifiers from a row type which forms a record.
+/// Ensure that ty is really a row type.
+fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+    ty.iter_as_rows()
+        .filter_map(|item| match item {
+            RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+            RowIteratorItem::Tail(..) => None,
+        })
+        .collect::<Vec<_>>()
+}
+
+/// Get the identifier before the `.`, for record completion.
+fn get_identifier(text: &str) -> Result<String, ResponseError> {
+    let name: String = text
+        .chars()
+        .rev()
+        .skip_while(|c| *c == '.') // skip . (if any)
+        .collect();
+
+    // unwrap is safe here because we know this is a correct regex.
+    let regex = regex::Regex::new(r"[_a-zA-Z0-9-']*[a-zA-Z]_?").unwrap();
```

Comment:
According to @matthew-healy in https://tweag.slack.com/archives/C0133ME1F4P/p1666780348629259?thread_ts=1666780145.455399&cid=C0133ME1F4P the regex used by the lexer is `_?[a-zA-Z][_a-zA-Z0-9-']*` which is different from the one here.

We must be consistent.

I would even go 1 step further and argue we should reuse the lexer's regexp rather than copying it here, if possible

Reply:
Yes, this one is the reverse of what we have in the lexer because the source text is reverse.

Reply:
For the reuse, I don't think constants literals are allowed for defining the regex in the lexer:
```rust
  #[regex("_?[a-zA-Z][_a-zA-Z0-9-']*")]
  Identifier(&'input str),
```

The `#regex` macro only accepts a string literal, I think. 

Even the lexer doesn't share common regexes: 
``` rust
 // **IMPORTANT**
// This regex should be kept in sync with the one for RawEnumTag below.
#[regex("_?[a-zA-Z][_a-zA-Z0-9-']*")]
Identifier(&'input str),

// **IMPORTANT**
// This regex should be kept in sync with the one for Identifier above.
#[regex("`_?[a-zA-Z][_a-zA-Z0-9-']*", |lex| lex.slice().split_at(1).1)]
RawEnumTag(&'input str),
```

Maybe the best we can do is add a comment also stating that the LSP should be updated as well, if there is any change to be made.

Reply:
Then we probably want to document this in both places, since this is a hidden domain coupling between both regexes: This regexp **must** be the reverse of the regexp in `<add link to lexer code here>` to detect/parse/identify identifiers

Also add a comment in the lexer regexp: any change in the lexer regex **must** also be backported in this regex.


We could probably go 1 step further but I suspect not everyone will agree: store regex components, store the list, and use the reverse list here:
```rust
// lexer.rs or whatever file contains the original regex :D
let optional_starting_underscore = r"_?"
let first_character = r"[a-zA-Z]"
let other_characters = r"[_a-zA-Z0-9-']*"
let pub identifier_regex_components = [optional_starting_underscore, first_character, other_characters]
let identifier_regex = regex::Regex::new(identifier_regex_components.join()).unwrap();

// completion.rs
use nickel::lexer::identifier_regex_components

/// Now the domain coupling between `lexer.rs` and this file is not only made explicit, but it's checked.
let reverse_identifier_regex = regex::Regex::new(identifier_regex_components.reverse().join()).unwrap();
```